libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Conway/Epoch.hs

```diff
@@ -26,8 +34,77 @@ newtype EpochExecEnv era = EpochExecEnv
 epochEnvSpec :: Specification fn (EpochExecEnv Conway)
 epochEnvSpec = TrueSpec
 
-epochStateSpec :: Specification fn (EpochState (ConwayEra StandardCrypto))
-epochStateSpec = TrueSpec
+epochStateSpec ::
+  Term ConwayFn EpochNo ->
+  Specification ConwayFn (EpochState (ConwayEra StandardCrypto))
+epochStateSpec epochNo = constrained $ \es ->
+  match es $ \_accountState ledgerState _snapShots _nonMyopic ->
+    match ledgerState $ \utxoState certState ->
+      match utxoState $ \_utxo _deposited _fees govState _stakeDistr _donation ->
+        match govState $ \ [var| proposals |] _committee constitution _curPParams _prevPParams _futPParams drepPulsingState ->
+          [ match constitution $ \_ policy ->
+              proposals `satisfies` proposalsSpec epochNo policy certState
+          , caseOn
+              drepPulsingState
+              -- DRPulsing
+              ( branch $ \pulser ->
+                  match pulser $ \_size _stakeMap _index _stakeDistr _stakePoolDistr _drepDistr _drepState pulseEpoch _committeeState _enactState pulseProposals _proposalDeposits _poolParams ->
+                    [ assert $ pulseEpoch ==. epochNo
+                    , forAll pulseProposals $ \gas ->
+                        match gas $ \gasId _ _ _ _ _ _ ->
+                          proposalExists gasId proposals
+                    , -- TODO: something is wrong in this case and I haven't figured out what yet
+                      assert False
+                    ]
+              )
+              -- DRComplete
+              ( branch $ \_snap ratifyState ->
+                  match ratifyState $ \enactState [var| enacted |] expired _delayed ->
+                    [ forAll expired $ \ [var| gasId |] ->
+                        proposalExists gasId proposals
+                    , -- TODO: this isn't enough, we need to ensure it's at most
+                      -- one of each type of action that's being enacted
+                      forAll enacted $ \govact ->
+                        [ reify proposals enactableProposals $ \enactable -> govact `elem_` enactable
+                        , assert $ not_ $ gasId_ govact `member_` expired
+                        ]
```

Comment:
Maybe you could use an `exists` here by generating some intermediate structure whose fields are `Maybe`'s of each proposal type where every `Just` contains a valid proposal of the corresponding type. Then make the proposals map a union of some invalid proposals and the proposals that are set to `Just` in that structure. Not sure if this will mess up the dependencies though.

Reply:
Yes, something like this makes sense

Reply:
Yes, something like this makes sense